### PR TITLE
Add ProjectLoadedInHost and PrioritizedProjectLoadedInHost

### PR DESCRIPTION
**Pull request by davkean**
This is the first part of fixing https://github.com/dotnet/project-system/issues/3414 by moving language service initialization to PrioritizedOnAfterOpenProject instead of OnAfterOpenProject.

Add two properties intended as a replacement for IProjectAsyncLoadDashboard usage. This has an added advantage of:

- Being able to hook onto both OnAfterOpenProject *and* PrioritizedOnAfterOpenProject solution events, instead of just OnAfterOpenProject.

- Avoid needing to combine IProjectAsyncLoadDashboard.ProjectLoadedInHost with IProjectAsynchronousTasksService.UnloadCancellationToken to avoid hanging when the project fails to open.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UnconfiguredProjectTasksService.cs [hunk 1]
```diff
@@ -3,20 +3,42 @@
...
+        [ProjectAutoLoad(completeBy:ProjectLoadCheckpoint.ProjectFactoryCompleted)]
+        [AppliesTo(ProjectCapability.CSharpOrVisualBasicOrFSharp)]
+        public Task OnProjectFactoryCompleted()
+        {
+            return _loadedInHostListener.StartListeningAsync();
+        }
+
+        public Task ProjectLoadedInHost
+        {
+            get { return _projectLoadedInHost.Task.WithCancellation(_tasksService.UnloadCancellationToken); }
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UnconfiguredProjectTasksService.cs]: Undecided whether to do this - *or* CancellationToken.Register. Was worried that the latter would get out of hand for large solutions as you can never unregister.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSolutionEventListener.cs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+            return HResult.OK;
+        }
+
+        public int PrioritizedOnAfterOpenProject(IVsHierarchy pHierarchy, int fAdded)
+        {
+            UnconfiguredProjectTasksService tasksService = GetUnconfiguredProjectTasksServiceIfApplicable(pHierarchy);
+            if (tasksService != null)
+            {
+                tasksService.OnPrioritizedProjectLoadedInHost();
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSolutionEventListener.cs]: Consider: `taskService?.OnPrioritizedProjectLoadedInHost();` (and above)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSolutionEventListener.cs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+            await _threadingService.SwitchToUIThread();
+
+            HResult result = _solution.Value.AdviseSolutionEvents(this, out _cookie);
+            if (result.Failed)
+                throw result.Exception;
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            await _threadingService.SwitchToUIThread();
```
Comment (jviau) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSolutionEventListener.cs]: Wrap this in `if (initialized) { ... }`. I think you will currently throw if disposed before initialization has finished.

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSolutionEventListener.cs]: Throw how?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSolutionEventListener.cs (hunk 2 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSolutionEventListener.cs]: Done.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSolutionEventListener.cs (hunk 3 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSolutionEventListener.cs]: I don't think we'll throw or can't think of a circumstance that we'll throw. We'll unnecessarily switch to the UI thread, so I've added a check.

Comment (jviau) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSolutionEventListener.cs]: Yeah I didn't notice the cookie is nil check. Thought you would unadvise without ever having advised in the first place.

**Comment by davkean** [comment]
Ended up merging this with #3447.

**Pull request closed without merging**